s3: add specific provider for Storj Shared gateways

elek opened the pull request:
<!--
Thank you very much for contributing code or documentation to rclone! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

#### What is the purpose of this change?

Make it easier (and documented) to use Storj distributed cloud storage via the s3 interface.

  1. Both on the Tardigrade and S3 backend it's clarified what are the pros and cons of using them.
  2. Storj added as a special s3 provider (with the predefined URL) 

#### Was the change discussed in an issue or in the forum before?

The discussion started here: https://forum.rclone.org/t/backward-compatibility-rules/26569/12
And later moved to #5616

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [ ] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

ncw reviewed (commented):
This is a great idea thanks.

I don't think the integration tests pass with this though do they?

Maybe we need to mask off the `Copy` feature flag and any other features which aren't implemented by Storj S3?

I'm not concerned about the file names which can't be uploaded.

elek commented:
> I don't think the integration tests pass with this though do they?

Good question. Let's record here the current state of integration test runs:

```
cd bacend/s3
go test -v -remote storjs3local: -fast-list -json | tee test2.json
jq '. | select(.Action=="fail") | .Test' test.json 

"TestIntegration/FsMkdir/FsEncoding/control_chars"
"TestIntegration/FsMkdir/FsEncoding/punctuation"
"TestIntegration/FsMkdir/FsEncoding/leading_LF"
"TestIntegration/FsMkdir/FsEncoding/leading_VT"
"TestIntegration/FsMkdir/FsEncoding/trailing_LF"
"TestIntegration/FsMkdir/FsEncoding/trailing_VT"
"TestIntegration/FsMkdir/FsEncoding"
"TestIntegration/FsMkdir/FsPutFiles/ObjectSetModTime"
"TestIntegration/FsMkdir/FsPutFiles/SetTier"
"TestIntegration/FsMkdir/FsPutFiles/FsPutStream/0"
"TestIntegration/FsMkdir/FsPutFiles/FsPutStream/100"
"TestIntegration/FsMkdir/FsPutFiles/FsPutStream"
"TestIntegration/FsMkdir/FsPutFiles"
"TestIntegration/FsMkdir/FsPutChunked/5Mi/5242880"
"TestIntegration/FsMkdir/FsPutChunked/5Mi/10485761"
"TestIntegration/FsMkdir/FsPutChunked/5Mi"
"TestIntegration/FsMkdir/FsPutChunked/5.054Mi/5300000"
"TestIntegration/FsMkdir/FsPutChunked/5.054Mi/10600001"
"TestIntegration/FsMkdir/FsPutChunked/5.054Mi"
"TestIntegration/FsMkdir/FsPutChunked/5.055Mi/5300053"
"TestIntegration/FsMkdir/FsPutChunked/5.055Mi/10600107"
"TestIntegration/FsMkdir/FsPutChunked/5.055Mi"
"TestIntegration/FsMkdir/FsPutChunked/6.667Mi/6990506"
"TestIntegration/FsMkdir/FsPutChunked/6.667Mi/13981013"
"TestIntegration/FsMkdir/FsPutChunked/6.667Mi"
"TestIntegration/FsMkdir/FsPutChunked/8Mi/8388608"
"TestIntegration/FsMkdir/FsPutChunked/8Mi/16777217"
"TestIntegration/FsMkdir/FsPutChunked/8Mi"
"TestIntegration/FsMkdir/FsPutChunked/10Mi/10485760"
"TestIntegration/FsMkdir/FsPutChunked/10Mi/20971521"
"TestIntegration/FsMkdir/FsPutChunked/10Mi"
"TestIntegration/FsMkdir/FsPutChunked"
"TestIntegration/FsMkdir"
"TestIntegration"
```

These are actually two problems:

 * Storj naming scheme is slightly more string (`FsEncoding`)
 * object metadata modifications are not supported in multipart upload. This is missing from the native protocol of Storj and tracked by storj/gateway-mt#39 Not an easy fix and this is responsible for all the FsPutChunked/FsPutFiles errors
 * ObjectSetModTime: It's the same for MPU will check if it requires a different fix for simple upload or not
 
 `SetTier` is not supported, this is the only one which can be turned off with features. I did it in 4daefaf294fc6c73176f7d003546eadcd3c7f9ad (thanks for the suggestion)

I don't see any copy test failed, but (just for the safety) I turned it off as well...

elek commented:
FTR: three encoding problems were fixed by eb3d8ff. `urlEncodeListings` is enabled for `StorJ` just like for minio.

Still some characters are not handled, though...

olefrost commented:
I also like the approach and addition of Storj to the list of S3 backends. 
 
I strongly recommend that the list of differences is expanded to give more details and facts. The current "use … to …" statements can easily be misunderstood and thereby lead to disappointed and frustrated rclone/Storj users.
 
The main issue is the implicit/hidden assumption, that Storj users have high-end (=pro) equipment with sufficiently free resources and internet bandwidth to honor the multipliers of the Stoj protocol without costs or delays. This makes the "use" statements wrong and misleading if the user has average equipment or network connection, as we found in the performance tests.
 
I therefore recommend you make the "use" statements clearer by explicitly stating the requirements and suggest you base them on [the conclusions from the performance tests](https://github.com/rclone/rclone/issues/5616#issuecomment-929553323): (with better language and formatting)

> … the S3 protocol is the best and simplest way to interface to StorJ for the ordinary rclone user (the default so to speak). 
>  
> The Tardigrade/StorJ protocol is interface for pro users having the equipment, internet connection (above 300/300 Mbit/s) and technical skills to configure and troubleshoot such an environment, e.g. increase the maximum number of open files/tcp connections on the system (requiring root/admin access and skills) - and of cause anybody with an interest in the distributed security model of Storj.
> 
 
I also recommend that the download multiplier is mentioned to complete the picture and help users find their own sweet spot.
 
I have noted that the multipliers you are using deviates from the [end user experiences shared in the Storj forum](https://forum.storj.io/t/amazon-s3-vs-tardigrade/4083). Do you know why? A guess: Storj strarts upload/download to 50% more servers than needed by the erasure multiplier and stops all when enough have completed.  It would be nice to have this clarified and adjusted in the docs, if needed.
 
I also recommend that any functional differences between the S3 and tardigrade interface are mentioned. For example, the difference in hash verification of uploads/downloads (it is apparently missing in the tardigrade backend [according to the docs](https://rclone.org/overview/#features))

I also recommend that any limitations in the Storj support of the S3 interface are mentioned, e.g. the currently missing support of the copy feature, discussed above.
 
I finally recommend that the section briefly mentions that client-side encryption also is possible with the S3 backend if combined the [rclone crypt backend](https://rclone.org/crypt/).

elek commented:
Thanks, @olefrost, for very good suggestions. 

I am not sure how is it possible in this favor of Markdown but I am thinking about a side-by-side comparison with a table where all of the mentioned differences can be summarized. And after that we can explain the details in text...

Will update the PR with it.

> increase the maximum number of open files/tcp connections on the system (requiring root/admin access and skills) -

I am not particularly convinced about this small detail. I think most linux environments have enough tcp/file limits to use the native connector. But I do agree to emphasize it as a difference 

> TCP connections to be used:
>     S3: a few (depends on parallelism)
>    multiple dozens to hundreds (separated connections to each data host)

olefrost commented:
> ... I am thinking about a side-by-side comparison with a table where all of the mentioned differences can be summarized. And after that we can explain the details in text...

Good idea!

> I think most linux environments have enough tcp/file limits to use the native connector.

Well I am not expert in linux or Storj, but the clone docs say this under [tardigrade known issues](https://rclone.org/tardigrade/#known-issues): 

> … If you do many transfers you eventually open a connection to most storage nodes (thousands of nodes).

My Ubuntu 20.04 LTS (on WSL2) has a process limit of 1024 and requires root access to change it. My Synology NAS has the same limit. My Windows pc has a limit of 3977 (from googling). I don’t know if my boxes are representative, but I have noted that several users in the forum are using surprisingly small equipment like pi4 and even home routers. That is why I found it important to highlight.

elek commented:
3d8470e7eac96ea11b1107064e114279a3d19f3b: more information about the differences (and suggestion about choosing). I decided to use a list format as I explained it in more detail, not just with +/- signs.

> I also recommend that any functional differences between the S3 and tardigrade interface are mentioned. For example, the difference in hash verification of uploads/downloads (it is apparently missing in the tardigrade backend according to the docs)

This is the only one, that I couldn't address. I don't know any special difference, after all both s3 gateway and tardigrade backend use the same uplink lib / protocols.

olefrost commented:
Great, that’s a very good overview!

I only have a few comments to improve layout and spelling:

Your table mark down doesn’t render correctly at my end (it actually doesn’t make a table at all :) The markdown syntax shown in this link however seems to work: [https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables)

Spelling: RCP -> TCP (in the overview), with combining -> when combining (in security)

>... functional differences between the S3 and tardigrade interface  ... This is the only one, that I couldn't address. I don't know any special difference, after all both s3 gateway and tardigrade backend use the same uplink lib / protocols.

I guess I was a bit to brief in my comment, let me elaborate a bit.

When I look in the Hash column in this backend feature table: [https://rclone.org/overview/#features](https://rclone.org/overview/#features) then I can see that the S3 backend supports MD5 hashes, so upload and downloads are verified when performed from/to backends supporting MD5 - and data integrity can be quickly verified with `rclone check`.

When I look at the tardigrade backend then it doesn’t support any hash, not even a custom hash like Dropbox. This means that uploads cannot be verified by rclone and I cannot use `rclone check` to quickly verify that the 10GB file on my computer is the same as the one in Storj. I will have to do a complete download and comparison using `rclone check --download` or similar.

This an important difference between the two backends and something I think should go into the table too.

elek commented:
> Your table markdown doesn’t render correctly at my end (it actually doesn’t make a table at all :)

Because it's not a table ;-) 

As I wrote I decided to make a list instead. The table is good if it only contains +/-, but I thought that table with a lot of text is less useful. But I can reformat, if you think it's better. And it's harder to maintain in markdown.

> When I look in the Hash column...

Thanks to explain it. I was not aware of this difference. I added an additional point with "limitations" where I mention it.

Typos are fixed, too (thanks to report)

olefrost commented:
> Because it's not a table ;-)

Ahh, sorry!

I played a little with the formatting and have found a layout that I find easier to read - and more in line with the rest of the documentation. Below is an example with the first two sections. I also did some rephrasing ;-) Take it as inspiration and  use what you like:

* Characteristics:
  * *Tardigrade backend*: Uses native RPC protocol, connects directly to the storage nodes which hosts the data. Requires more CPU resource of encoding/decoding and has network amplification (especially during the upload), uses lots of TCP connections 
  * *S3 backend*: Uses S3 compatible HTTP Rest API via the shared gateways. There is no network amplification, but performance depends on the shared gateways and the secret encryption key should be shared with the gateway. 
* Typical usage:
  * *Tardigrade backend*: Server environments and desktops with enough resources, internet speed and connectivity - and applications where tardigrades client-side encryption is required.
  * *S3 backend*: Desktops and similar with limited  resources, internet speed or connectivity.

here is the raw text (without linebreaks):

```
* Characteristics:
  * *Tardigrade backend*: Uses native RPC protocol, connects directly to the storage nodes which hosts the data. Requires more CPU resource of encoding/decoding and has network amplification (especially during the upload), uses lots of TCP connections 
  * *S3 backend*: Uses S3 compatible HTTP Rest API via the shared gateways. There is no network amplification, but performance depends on the shared gateways and the secret encryption key should be shared with the gateway. 
* Typical usage:
  * *Tardigrade backend*: Server environments and desktops with enough resources, internet speed and connectivity - and applications where tardigrades client-side encryption is required.
  * *S3 backend*: Desktops and similar with limited resources, internet speed or connectivity.
```

Typo: Tadigrade -> Tardigrade (several places)

elek commented:
Thanks for the suggestions.

I agree with your phrasing, I just started to use it (almost) as is ;-) 

I did only one micro modification: using `end-to-end` encryption instead of `client-side` encryption. (I think it's more expressive...)

Also fixed the typos and started to use multi-level list as formatting:

![image](https://user-images.githubusercontent.com/170549/136182378-9b405f05-d159-4d8c-b4f9-c45099e2c7ca.png)

olefrost commented:
Thanks, looks good to me!

> I did only one micro modification: using end-to-end encryption instead of client-side encryption. (I think it's more expressive...)

That is OK with me, I can see that it is often used by Storj e.g. in this blog post:
https://www.storj.io/blog/what-is-end-to-end-encryption

I do however find the term end-to-end encryption weaker than client-side encryption when it comes to cloud storage.

Example: I have end-to-end encryption to my GoogleDrive (because the REST API uses HTTPS/TLS), but this does not keep my data safe from preying eyes at Google. If I use the crypt backend, Cryptomator or something similar on top of my Google Drive, then I have client-side encryption and is protected from preying eyes at Google.

This is in line with the definitions found on Wikipedia:
https://en.wikipedia.org/wiki/End-to-end_encryption
https://en.wikipedia.org/wiki/Client-side_encryption

elek commented:
Ok, I am convinced (3888f7a).

elek commented:
Hi, thanks for all the feedback.

Please let me know if need to improve anything here.

>  All commit messages are in house style.

Do I need to squash commits to one or it's done during the merge?

elek commented:
I would suggest separating the question of 64 MB chunks from this generic doc update + providers.

As far as I know, there is no open question here, but please let me know if I need to update it in any way...

Thanks a lot...

ncw commented:
Yesterday I merged a patch which deals with the differences between the providers in a more sane way.

Can you rebase this on master, then add a `Storj` case to `setQuirks` and work out what quirks (if any) Storj needs? You can test with the integration tests (use `go test -v -remote rackcorp:` in the backend/s3 directory).

https://github.com/rclone/rclone/blob/25366268fe921608ee68c6dc33b744b80d860abe/backend/s3/s3.go#L1692-L1698

The way to change the min chunk size would be to adjust it in setQuirks to 64MB as a minimum.

Thank you :-)

elek commented:
Thanks for the hint @ncw, I created the `setQuirck` case for `Storj` and set the chunk size (64MB). Other quirks are not necessary (fine with `urlEncodeListings = true`).

Rebased the branch and tested it from CLI with creating bucket / uploading data to Storj gatewayt...

elek commented:
@ncw are you fine with this approach? Would be happy to further improve this PR if you have any suggestions...

elek commented:
Hi, can anybody help this PR to be merged, please? If you need any more modifications, please let me know...

calebcase reviewed (approved):
LGTM Thanks!

calebcase commented:
@ncw Anything else we should adjust?

elek commented:
Hi, please let me know if I need to improve this patch in any way...

olefrost commented:
Hi @elek, I haven’t seen any @ncw online activity since December 9th; I hope he is OK and returns.

Tip: The [rclone commit log](https://github.com/rclone/rclone/commits/master) can be used to monitor approval/merge activity.

ncw commented:
@elek I've squashed, rebased and reworded a few docs so I think this is ready to merge if you are OK?

elek commented:
> @elek I've squashed, rebased and reworded a few docs so I think this is ready to merge if you are OK?

Sure, I am fine with that. Thank you very much.

ncw commented:
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.58

Thank you for taking a look @elek
